docs/concepts/templating.md

```diff
@@ -0,0 +1,298 @@
+that only they care about (for instance, the Footer can gather all the copyright links itself, those will always be the same). But for things like, say, Page Title, 
+the component needs to be able to render whatever you pass it, since it'll be different on every page or post. That's a great place for the Controller to step in - 
+it's aware of the context (remember, we called this `loop` method specifically from our `index.php`, so we _know_ we're in the Index Loop). So your Controller should 
+make those decisions and gather any data specific to its own context. 
+
+Before we move on, let's take a quick peek at one of the methods being called here to see what types of data we're going to be passing to our Document component. 
+Let's look at `get_footer_content()`. 
+
+```php
+protected function get_footer_content() {
+    $menu = new Menu();
+
+    return [
+        'navigation' => [
+            'menu' => $menu->secondary(),
+        ],
+    ];
+}
+```
+
+Ooooh now _that's_ interesting. We're not returning a string to be rendered, but an array with its own arguments. Why would we do that? Have we lost our minds??
+
+Yes, but this also has a good reason: the array of data we're returning is going to be used by the Document component to call, in turn, the Footer component. Since it's
+rendering a component, and not just echoing something, we need to pass the _arguments that that component needs in order to function_. 
+
+This is, perhaps, the biggest hill to climb in our entire Templating system. Since Components compose other Components, we need to ensure that the data we pass to them to
+_do_ that rendering is an array of arguments for that component. It can be a little tricky to wrap your head around, but the basic rule of thumb is this: If your component
+is composing others, you're going to need to pass it an array of arguments for that component. Anything it simply echos can be passed as a string (or in the case of classes 
+and attributes, an array of values). 
+
+### Step 3: It's Component Time!
+
+Now that we've gathered our top-level data and kicked off the render of the Document component, we live entirely in Component land! From here, everything that renders, 
+from page content to post lists to buttons to sidebars, is a component. We no longer need to call other Controllers or worry about any other routing. We've started the Components 
+train, and all we have to do is hop aboard!
+
+Let's look at the Document component:
+
+```php
+class Document extends Component {
+
+	public function init() {
+		$this->data['language_attributes'] = $this->get_language_attributes();
+		$this->data['body_class']          = $this->get_body_class();
+	}
+
+	protected function get_language_attributes() {
+		ob_start();
+		language_attributes();
+
+		return ob_get_clean();
+	}
+
+	protected function get_body_class() {
+		return implode( ' ', get_body_class() );
+	}
+
+	public function render(): void {
```

Comment:
The biggest thing that gives me pause about this new architecture is this. Adding twig in php. Did you give any thoughts to doing something like `echo $this-render('my/template/index.twig', $this-data);`. It's a bit closer to other MVC frameworks and very similar to `$this->render_component( 'document/Document.php', $args );`

Reply:
So we went back and forth on this about 10 times. I think every day when I wake up I have a different opinion. The advantage of having the Twig in the PHP is that it's all self-contained, and there's no juggling of files to get your view up while working with your component data. It basically acts like a cleaner, more-structured version of how template-parts work in most WP themes. 

However, there are some disadvantages, such as Twig autocompleting/typehinting not working in most IDEs, and of course just the sort of weirdness of having markup in your PHP. 

I think I could go either way, and I'm happy to go with the approach that the majority here decides on. It's relatively minor to just pull the Twig out to it's own file once again and call it from within the Component, so I'm perfectly happy doing so if we decide that's the best route!

Reply:
Follow-up on this: I'm having Sam and Ryan try out the separate-Twig-file approach as they work on some Blocks to see how it all feels. I'll use that as a good determining factor on whether it makes more sense to separate things out.Stay tuned!

Reply:
I _think_ Ryan, Sam and I have landed on keeping separate twig files rather than rendering twig inside PHP. The pain of IDE syntax intellisense alone is enough to say this is probably not worth it.